File: VB/WinWebSolution.Module/ModelExtenders.vb
Imports Microsoft.VisualBasic
Imports System.ComponentModel
Imports DevExpress.ExpressApp.DC
Imports DevExpress.ExpressApp.Model

Namespace WinWebSolution.Module

	Public Interface IModelClassAllowSort
		<Category("Behavior")> _
		Property DefaultListViewAllowSort() As Boolean
	End Interface
	Public Interface IModelMemberAllowSort
		<Category("Behavior")> _
		Property AllowSort() As Boolean
	End Interface
	Public Interface IModelListViewAllowSort
		<Category("Behavior")> _
		Property AllowSort() As Boolean
	End Interface
	Public Interface IModelColumnAllowSort
		<Category("Behavior")> _
		Property AllowSort() As Boolean
	End Interface
	<DomainLogic(GetType(IModelClassAllowSort))> _
	Public NotInheritable Class ModelClassAllowSortLogic
		Public Shared Function Get_DefaultListViewAllowSort(ByVal modelClass As IModelClass) As Boolean
			If modelClass IsNot Nothing AndAlso modelClass.TypeInfo IsNot Nothing Then
				Dim attribute As AllowSortAttribute = modelClass.TypeInfo.FindAttribute(Of AllowSortAttribute)()
				If attribute IsNot Nothing Then
					Return attribute.AllowSort
				End If
			End If
			Return True
		End Function
	End Class
	<DomainLogic(GetType(IModelMemberAllowSort))> _
	Public NotInheritable Class ModelMemberAllowSortLogic
		Public Shared Function Get_AllowSort(ByVal modelMember As IModelMember) As Boolean
			If modelMember IsNot Nothing AndAlso modelMember.MemberInfo IsNot Nothing Then
				Dim attribute As AllowSortAttribute = modelMember.MemberInfo.FindAttribute(Of AllowSortAttribute)()
				If attribute IsNot Nothing Then
					Return attribute.AllowSort
				End If
			End If
			Return True
		End Function
	End Class
	<DomainLogic(GetType(IModelListViewAllowSort))> _
	Public NotInheritable Class ModelListViewAllowSortLogic
		Public Shared Function Get_AllowSort(ByVal modelListView As IModelListView) As Boolean
			If modelListView IsNot Nothing AndAlso modelListView.ModelClass IsNot Nothing Then
				Return (CType(modelListView.ModelClass, IModelClassAllowSort)).DefaultListViewAllowSort
			Else
				Return True
			End If
		End Function
	End Class
	<DomainLogic(GetType(IModelColumnAllowSort))> _
	Public NotInheritable Class ModelColumnAllowSortLogic
		Public Shared Function Get_AllowSort(ByVal modelColumn As IModelColumn) As Boolean
			If modelColumn IsNot Nothing AndAlso modelColumn.ModelMember IsNot Nothing Then
				Return (CType(modelColumn.ModelMember, IModelMemberAllowSort)).AllowSort
			Else
				Return True
			End If
		End Function
	End Class
End Namespace

File: CS/WinWebSolution.Module/ModelExtenders.cs
using System.ComponentModel;
using DevExpress.ExpressApp.DC;
using DevExpress.ExpressApp.Model;

namespace WinWebSolution.Module {

    public interface IModelClassAllowSort {
        [Category("Behavior")]
        bool DefaultListViewAllowSort { get; set; }
    }
    public interface IModelMemberAllowSort {
        [Category("Behavior")]
        bool AllowSort { get; set; }
    }
    public interface IModelListViewAllowSort {
        [Category("Behavior")]
        bool AllowSort { get; set; }
    }
    public interface IModelColumnAllowSort {
        [Category("Behavior")]
        bool AllowSort { get; set; }
    }

    // Logic

    [DomainLogic(typeof(IModelClassAllowSort))]
    public static class ModelClassAllowSortLogic {
        public static bool Get_DefaultListViewAllowSort(IModelClass modelClass) {
            AllowSortAttribute attribute = modelClass.TypeInfo.FindAttribute<AllowSortAttribute>();
            if (attribute != null) {
                return attribute.AllowSort;
            } else {
                return true;
            }
        }
    }
    [DomainLogic(typeof(IModelMemberAllowSort))]
    public static class ModelMemberAllowSortLogic {
        public static bool Get_AllowSort(IModelMember modelMember) {
            AllowSortAttribute attribute = modelMember.MemberInfo.FindAttribute<AllowSortAttribute>();
            if (attribute != null) {
                return attribute.AllowSort;
            } else {
                return true;
            }
        }
    }
    [DomainLogic(typeof(IModelListViewAllowSort))]
    public static class ModelListViewAllowSortLogic {
        public static bool Get_AllowSort(IModelListView modelListView) {
            return ((IModelClassAllowSort)modelListView.ModelClass).DefaultListViewAllowSort;
        }
    }
    [DomainLogic(typeof(IModelColumnAllowSort))]
    public static class ModelColumnAllowSortLogic {
        public static bool Get_AllowSort(IModelColumn modelColumn) {
            return ((IModelMemberAllowSort)modelColumn.ModelMember).AllowSort;
        }
    }
}